mapper-processor/src/test/java/com/datastax/oss/driver/internal/mapper/processor/util/HierarchyScannerTest.java

```diff
@@ -0,0 +1,208 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.datastax.oss.driver.internal.mapper.processor.util;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import com.datastax.oss.driver.api.mapper.annotations.HierarchyScanStrategy;
+import com.datastax.oss.driver.shaded.guava.common.collect.Lists;
+import java.util.Arrays;
+import java.util.List;
+import javax.lang.model.element.Name;
+import javax.lang.model.element.TypeElement;
+import javax.lang.model.type.TypeKind;
+import javax.lang.model.type.TypeMirror;
+import javax.lang.model.util.Types;
+import org.junit.Test;
+import org.mockito.Mockito;
+
+public class HierarchyScannerTest {
+
+  Types types = Mockito.mock(Types.class);
+
+  @Test
+  public void should_build_proper_hierarchy_with_default_strategy() {
+    /*
+     * given the following hierarchy
+     *
+     *                  a
+     *                 /
+     *                z
+     *               /
+     *         a    y
+     *        /    /
+     *        b   x w
+     *         \ / /
+     *          c
+     */
+    MockInterface a = i("a");
+    MockInterface z = i("z", a);
+    MockInterface y = i("y", z);
+    MockInterface x = i("x", y);
+    MockInterface w = i("w");
+    MockClass b = c("b", null, a);
+    MockClass c = c("c", b, x, w);
+
+    // when no HierarchyScanStrategy is defined, then the default behavior should be used
+    // which is to scan the entire tree and return it in the correct (bottom up) order.
+    assertThat(HierarchyScanner.resolveTypeHierarchy(c.classElement, types).toString())
+        .isEqualTo("[c, b, x, w, a, y, z]");
+  }
+
+  @Test
+  @SuppressWarnings("unchecked")
+  public void should_build_property_hierarchy_with_strategy_defined() {
+    /*
+     * given the following hierarchy
+     *
+     *                  a
+     *                 /
+     *        a r     z
+     *       / /     /
+     *      b       y
+     *       \     /
+     *       @d   x w
+     *         \ / /
+     *          c
+     */
+    // with a HierarchyScanStrategy annotation on "d" that dictates that "b" is the highest class,
+    // but to not include it.
+    HierarchyScanStrategy bHighest = Mockito.mock(HierarchyScanStrategy.class);
+    // return this class's name, which will also be used for  b's class name.
+    Mockito.when(bHighest.highestAncestor()).thenReturn((Class) this.getClass());
+    Mockito.when(bHighest.includeHighestAncestor()).thenReturn(false);
+    MockInterface a = i("a");
+    MockInterface r = i("r");
+    MockInterface z = i("z", a);
+    MockInterface y = i("y", z);
+    MockInterface x = i("x", y);
+    MockInterface w = i("w");
+
+    MockClass b = c(this.getClass().getName(), null, a, r);
```

Comment:
is there a particular reason why in this mock we choose to use `this.getClass().getName()` instead of the following pattern from different c mocks? (like `c("b", null, a,r)`
Same question to line https://github.com/datastax/java-driver/pull/1250/files#diff-26867dfb0fa8ecefa97f1251414342dcR85

Reply:
We need to do it this way because `HierarchyScanStrategy.highestAncestor` takes a `Class` and you cannot mock classes since they are final.

Reply:
with you suggestion to use `Classes.isSame()` I no longer need to do this :+1: